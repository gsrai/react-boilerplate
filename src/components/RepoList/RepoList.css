.repo-list__heading {
  text-align: center;
}