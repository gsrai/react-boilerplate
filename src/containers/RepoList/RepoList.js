import React from 'react'
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import { actions as repoListActions } from '../../reducers/RepoListReducer'
import RepoListItem from '../../components/RepoListItem'

class RepoList extends React.Component {
  componentDidMount() {
    this.props.actions.loadRepoListInit()
    this.props.actions.loadRepoList('preact')
  }

  render() {
    const { loading, loadSuccess, loadError, reason, repos } = this.props
    return (
      <div>
        <h1>Example</h1>
        <div>
          { loading &&
            'loading repositories...'
          }
          { loadSuccess &&
            repos.map(repo => (<RepoListItem repo={repo} />))
          }
          { loadError &&
            'A problem has occurred: ' + reason
          }
        </div>
      </div>
    )
  }
}

RepoList.propTypes = {
  loading: PropTypes.bool.isRequired,
  loadSuccess: PropTypes.bool.isRequired,
  loadError: PropTypes.bool.isRequired,
  reason: PropTypes.string,
  repos: PropTypes.array,
  actions: PropTypes.object.isRequired
}

export default connect(
  state => {
    const loading = state.repoListReducer.loading
    const loadSuccess = state.repoListReducer.loadSuccess
    const loadError = state.repoListReducer.loadError
    const reason = state.repoListReducer.reason

    const repos = state.repoListReducer.repos

    return { loading, loadSuccess, loadError, reason, repos }
  },
  dispatch => ({
    actions: bindActionCreators(repoListActions, dispatch)
  })
)(RepoList)
