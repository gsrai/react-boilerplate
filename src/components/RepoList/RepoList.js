import React from 'react'
import PropTypes from 'prop-types'
import RepoListItem from '../RepoListItem'

import './RepoList.css'

export const RepoList = (props) => {
  const { loading, loadSuccess, loadError, reason, repos } = props
  return (
    <div>
      <h1 className='repo-list__heading'>Example</h1>
      <div>
        { loading &&
          'loading repositories...'
        }
        { loadSuccess &&
          repos.map((repo, i) => (<RepoListItem key={i} repo={repo} />))
        }
        { loadError &&
          'A problem has occurred: ' + reason
        }
      </div>
    </div>
  )
}

RepoList.propTypes = {
  loading: PropTypes.bool.isRequired,
  loadSuccess: PropTypes.bool.isRequired,
  loadError: PropTypes.bool.isRequired,
  reason: PropTypes.string,
  repos: PropTypes.array
}

export default RepoList
