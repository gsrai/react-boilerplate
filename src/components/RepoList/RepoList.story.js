import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import RepoList from './RepoList'

const generateProps = (opts) => {
  return {
    loading: opts.loading || false,
    loadSuccess: opts.loadSuccess || false,
    loadError: opts.loadError || false,
    reason: opts.reason || '',
    repos: opts.repos || [],
    loadRepos: opts.loadRepos || action('clicked')
  }
}

const mockRepos = [
  {
    full_name: 'facebook/react',
    html_url: 'https://github.com/facebook/react',
    description: 'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
    stargazers_count: 108963
  }
]

storiesOf('Components / RepoList', module)
  .add('default', () => <RepoList {...generateProps({})} />)
  .add('loading', () => <RepoList {...generateProps({ loading: true })} />)
  .add('with Repos loaded', () => <RepoList {...generateProps({ loadSuccess: true, repos: mockRepos })} />)
  .add('with loading error', () => <RepoList {...generateProps({ loadError: true, reason: 'Network Error' })} />)
