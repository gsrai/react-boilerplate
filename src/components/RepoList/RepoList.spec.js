import { shallow } from 'enzyme'
import React from 'react'
import RepoList from './RepoList'

describe('foo', () => {
  test('component should render loading text when loading', () => {
    const props = genProps({ loading: true })
    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should render repo data when successfully loaded', () => {
    const repoData = [
      {
        full_name: 'facebook/react',
        html_url: 'https://github.com/facebook/react',
        description: 'A declarative, efficient, and flexible JavaScript library for building user interfaces.',
        stargazers_count: 108963
      }
    ]
    const props = genProps({ loadSuccess: true, repos: repoData })
    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('component should pass reason when there is an error', () => {
    const reason = 'cannot load repos'
    const props = genProps({ loadError: true, reason })
    const wrapper = shallow(<RepoList {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

// maybe this should be a fixture/s
function genProps(opts) {
  return {
    loading: opts.loading || false,
    loadSuccess: opts.loadSuccess || false,
    loadError: opts.loadError || false,
    reason: opts.reason || '',
    repos: opts.repos || []
  }
}
