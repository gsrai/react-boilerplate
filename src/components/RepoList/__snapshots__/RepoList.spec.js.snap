// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foo component should pass reason when there is an error 1`] = `
<div>
  <h1
    className="repo-list__heading"
  >
    Example
  </h1>
  <div>
    A problem has occurred: cannot load repos
  </div>
</div>
`;

exports[`foo component should render loading text when loading 1`] = `
<div>
  <h1
    className="repo-list__heading"
  >
    Example
  </h1>
  <div>
    loading repositories...
  </div>
</div>
`;

exports[`foo component should render repo data when successfully loaded 1`] = `
<div>
  <h1
    className="repo-list__heading"
  >
    Example
  </h1>
  <div>
    <RepoListItem
      key="0"
      repo={
        Object {
          "description": "A declarative, efficient, and flexible JavaScript library for building user interfaces.",
          "full_name": "facebook/react",
          "html_url": "https://github.com/facebook/react",
          "stargazers_count": 108963,
        }
      }
    />
  </div>
</div>
`;
